Port competition homepage to rails

timhabermaas opened the pull request:
Continues the work from #407, mainly adds some performance improvements.

Todo
---

* [x] Get rid of floatThead, because it is buggy and slow
* [x] Fix specs
* [x] Add simple feature test to catch stupid errors

timhabermaas commented on WcaOnRails/app/controllers/registrations_controller.rb:
```diff
@@ -45,29 +45,29 @@ def psych_sheet
     @competition = competition_from_params
     @event = Event.find(params[:event_id])
-    preferred_format = @event.preferred_formats.first
+    @preferred_format = @event.preferred_formats.first
```
@jfly This caused me some headaches because stuff broke after rebasing with `master`. Could you take a look at this commit: 405c27e361288b7c994b6bdc2e9fb674159118b6 and check whether I messed up some stuff.

btw. Was the renaming `position` to `pos` really necessary? It's gonna cause more conflicts unless I'll quickly merge this branch into master.

timhabermaas commented:
With `floatThead`:

<img width="287" alt="screen shot 2016-04-13 at 2 50 09 pm" src="https://cloud.githubusercontent.com/assets/96054/14493738/31c8c2b4-0187-11e6-8dad-1f8503c7a467.png">

Without `floatThead`:

<img width="276" alt="screen shot 2016-04-13 at 2 51 00 pm" src="https://cloud.githubusercontent.com/assets/96054/14493741/3402d830-0187-11e6-8b07-10c7194ecc0a.png">

timhabermaas commented:
@jfly I think I'm finished. Could someone help me get this branch into the staging environment in order to test performance? (@pedrosino?)

pedrosino commented:
Moved it to staging. Check
https://staging.worldcubeassociation.org/competitions/FMCEurope2016

Looking good :)

On Wed, Apr 13, 2016 at 6:04 PM, Tim Habermaas <notifications@github.com>
wrote:

> @jfly <https://github.com/jfly> I think I'm finished. Could someone help
> me get this branch into the staging environment in order to test
> performance? (@pedrosino <https://github.com/pedrosino>?)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cubing/worldcubeassociation.org/pull/485#issuecomment-209646734>
>

pedrosino commented:
One thing I don't particularly love, but could leave with, is that we right-align the stuff inside col-md-6. This creates a big gap between Information on the right and the rest, and also the space doesn't look well used (compare the blue alert with the rest)
![485](https://cloud.githubusercontent.com/assets/1881933/14510096/12f3c720-01a6-11e6-9abf-5621c5585c33.png)

timhabermaas commented:
Thanks a lot, @pedrosino!

I had to run the migrations (competitions with results require the `updated_at` column as cache key), but it seems to run flawlessly now. The first request to a big competition (e.g. USNationals2015) takes about 2.5s and the cached version of it takes about 200ms. I'm pretty pleased with that.

> This creates a big gap between Information on the right and the rest, and also the space doesn't look well used

How would you structure it? Maybe we could move the delegate/organizer to the right? I don't really have any preferences.

pedrosino commented:
Maybe left-align everything and make it col-md-4 instead?
Moving something to the right column also would make Information look less lonely.

jonatanklosko commented on WcaOnRails/app/helpers/application_helper.rb:
```diff
@@ -97,6 +111,128 @@ def wca_table(responsive: true, hover: true, striped: true, table_class: "", dat
+      end
+      block.call(table)
+    end
+  end
```
selectable-rows.js has been removed in #455 and now we use bootstrap-table

timhabermaas replied:
Thanks, I've removed it (and `wca_table_for`).

jfly commented on WcaOnRails/app/models/light_result.rb:
```diff
@@ -0,0 +1,107 @@
+require 'solve_time'
+require_relative './format'
+
+class LightResult
```
You have a very nice description of why this class exists here: https://github.com/timhabermaas/worldcubeassociation.org/commit/511ac52955bcdeac5fcb3095e9e3afbd6a6eec7e. Could you add that description to this class?

timhabermaas replied:
Yes, sure.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -458,11 +465,82 @@ def in_progress?
+  end
+
+  def results_updated_at
+    Option.from_nilable(results.order('updated_at desc').pluck(:updated_at).first)
```
I don't fully understand this Option class yet, so I might be missing something, but `results.order('updated_at desc').first.updated_at` seems simpler to me thank `pluck(...).first`.

timhabermaas replied:
They behave the same way, but `pluck` doesn't load the entire `ActiveRecord` object and also only `SELECT`s that particular column, see http://apidock.com/rails/ActiveRecord/Calculations/pluck.

jfly replied:
Ahh, I see. But won't a `.pluck` cause a select of every single result and then generate a ruby array of their `updated_at` fields? That feels worse than selecting every column of exactly 1 row. See:

```
irb(main):004:0> r = Result.pluck(:id); nil
   (1.4ms)  SELECT `Results`.`id` FROM `Results`
D, [2016-04-15T02:30:42.953473 #12480] DEBUG -- :    (1.4ms)  SELECT `Results`.`id` FROM `Results`
=> nil
irb(main):006:0> r = Result.first.id; nil
  Result Load (0.4ms)  SELECT  `Results`.* FROM `Results`  ORDER BY `Results`.`id` ASC LIMIT 1
D, [2016-04-15T02:31:01.408759 #12480] DEBUG -- :   Result Load (0.4ms)  SELECT  `Results`.* FROM `Results`  ORDER BY `Results`.`id` ASC LIMIT 1
=> nil
```

Perhaps the best of both worlds?

```
irb(main):005:0> r = Result.limit(1).pluck(:id); nil
   (0.3ms)  SELECT  `Results`.`id` FROM `Results` LIMIT 1
D, [2016-04-15T02:30:53.736865 #12480] DEBUG -- :    (0.3ms)  SELECT  `Results`.`id` FROM `Results` LIMIT 1
=> nil
```

timhabermaas replied:
Oh god, I'm stupid... I'll fix this, thanks!

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -458,11 +465,82 @@ def in_progress?
+  end
+
+  def result_cache_key(view)
+    [id, view, results_updated_at.map(&:iso8601).unwrap_or("")]
```
I'm having trouble understanding the motivation of this Option class. This feels like an idiom that you are familiar with that I've never seen before. Is the idea to deal with Competitions that don't have results without having to write an explicit if statement for them?

timhabermaas replied:
Instead of encoding "there is no results_updated_at date" by returning `nil` and returning a `DateTime` otherwise, I return `None` and `Some.new(...)`. I'm a big fan of method's return values having the same exact interface. That's not the case with `nil`. So, it's mainly to avoid `nil` creeping in since you can't call the methods you want to call without explicitly _unwrapping_ the value first. Having methods like `map` on `Option` also avoids `if`, but that wasn't the main motivation.

The wikipedia article has some more information: https://en.wikipedia.org/wiki/Option_type and I'm borrowing Rust's names: https://doc.rust-lang.org/std/option/enum.Option.html

If the concept is too alien for anyone, I'm going back to returning `nil` from `results_updated_at` even though I'd consider that broken. :see_no_evil:

jfly replied:
Interesting. That sounds like it could be a useful idiom once you're used to it. Thanks for the details.

What about avoiding this problem entirely? In this particular case, if there are no Results for this competition, we could use the competition's updated_at as a cache key. Oh wait, Competitions don't have an updated_at field... What about adding an `updated_at` field, or using a Competition's start date?

timhabermaas replied:
This sounds like a good idea, I think I'll do that.

timhabermaas replied:
I just removed `results_updated_at` completely because falling back to `start_date` doesn't make much sense.

jfly commented on WcaOnRails/lib/solve_time.rb:
```diff
@@ -1,21 +1,65 @@
+    wca_value == SKIPPED_VALUE
+  end
+
+  def to_orderable
```
I'm pretty surprised at the ordering here. Don't we want worse solves to be treated as bigger? (DNF is worse than a 10 second solve, ie: bigger) I feel like I'd change all of this to `self.skipped? ? 1 : 0`...

timhabermaas replied:
I've fixed this, but I had to keep the `dnf?` case in here since DNF is represented as `-1` which would lead to `DNF < solve_time(30)`.

jfly commented on WcaOnRails/spec/models/solve_time_spec.rb:
```diff
@@ -0,0 +1,48 @@
+      expect(SolveTime::DNS < solve_time(30)).to eq true
+    end
+
+    it "treats DNS as worst than a finished solve" do
```
nit: worst -> worse

jfly commented on WcaOnRails/spec/models/solve_time_spec.rb:
```diff
@@ -0,0 +1,48 @@
+    end
+
+    it "treats DNS as worst than a finished solve" do
+      expect(SolveTime::DNF < solve_time(30)).to eq true
```
As I mentioned in `to_orderable` in SolveTime, I'm surprised at this ordering... I would expect DNF to be greater than 30.

jfly commented:
This looks really great, @timhabermaas. I'm going to a competition this weekend, but I'll try to be responsive so we can get this merged up quickly.

jfly commented on WcaOnRails/spec/models/solve_time_spec.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    it "treats skipped as worst" do
+      expect(SolveTime::SKIPPED < SolveTime::DNF).to eq true
```
Since skipped is supposed to be worst, shouldn't the less than here actually be a greater than?

timhabermaas replied:
I don't know, you wrote that test. :P

But it makes sense, I'll fix this. Btw. I don't think we compare skipped solves right now since we `reject` skipped solves: https://github.com/cubing/worldcubeassociation.org/pull/485/files#diff-18cb788097862a70cf83de74fd4c2488R82 before sorting.

jfly replied:
Oh, weird. No idea what I was thinking when I wrote that.

jfly commented on WcaOnRails/spec/models/solve_time_spec.rb:
```diff
@@ -0,0 +1,50 @@
+    end
+
+    it "treats DNS as worse than DNF" do
+      expect(SolveTime::DNS < SolveTime::DNF).to eq true
```
Same here, worse means greater than, right?

timhabermaas commented:
@jfly I've rebased on top of `master` and fixed the rubocop errors which came up.

If you find the time, feel free to merge it. There's still one unresolved comment from @pedrosino (https://github.com/cubing/worldcubeassociation.org/pull/485#issuecomment-209655199) which I think can wait until after the merge. You can clearly read the information and I can't be trusted with design decisions. I'm also kinda sick of resolving conflicts. :hamster:

jfly commented:
Thanks, Tim! Merging up now.

mkoryak commented:
floatThead taking 32 seconds to do "something" is insane. can you point me to the url where this was happening.

timhabermaas commented:
> floatThead taking 32 seconds to do "something" is insane. can you point me to the url where this was happening.

floatThead didn't make it into production, but this is the page which I measured the rendering performance on: https://www.worldcubeassociation.org/competitions/USNationals2015/results/by_person If you want to reproduce it you can add the class `floatThead` to each `<table>`.

> 90% of the slowness comes from the browser having to reflow the table layout, which can be slow with complicated tables.
> 
> https://kangax.github.io/compat-table/es6/
> That is probably the most complicated/giant table where the plugin runs (that I know of) and it doesn't take a half of a minute to load.

Yes, that's much quicker than what I experienced. I tried using it on a page with ~500 tables, though. This might have something to do with it. We use floatThead on other pages and it works fine.

mkoryak commented:
oh, that makes sense then. I think you could make it work by adding a scroll event listener that initializes floatThead on which are in the viewport, that way the plugin doesnt need to run on 500 tables at once.

You might also need to destroy it on tables which are no longer in viewport (if the page starts to slow down as you scroll past more and more tables)

timhabermaas commented:
Thanks for the tips, @mkoryak! We'll see if we can make it work.

mkoryak commented:
btw, where on your site do you use the plugin? I wanted to link it from the readme, but your site is pretty big, I couldn't find it.

timhabermaas commented:
We use it here: https://www.worldcubeassociation.org/competitions/Euro2016/registrations I'm pretty sure there are more usages, but I can't remember/find them right now.

mkoryak commented:
Thanks, 
also, I didn't realize that you guys are crazy. "Rubik's Cube: With feet" 
o_0
